### PR TITLE
doc: Update threat model to treat untrusted upstreams and downstreams equally

### LINKED ISSUE #12278: Harden Envoy against untrusted upstream servers
This is top level issue for tracking tasks and progress of hardening Envoy against untrusted upstream servers. More specific sub-tasks are split into its own issues and linked to this issue.

### Objective

1. Fix existing known high priority issues with untrusted upstream servers, sufficient to allow Envoy core components to be hardened against untrusted upstream servers.
1. Modify the Envoy OSS threat model to include untrusted upstream servers.

### Background
Presently [Envoy’s threat model](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/threat_model) considers all upstream data and control plane infrastructure to be trusted. This includes upstream servers, auxiliary servers invoked by filters, such as remote authz server, control plane endpoints or user controlled portions of configuration. This approach meets requirements of all existing Envoy’s production deployments which are mostly reverse proxy to trusted production services or service mesh. As a result most security related fixes and improvements have been applied to client (downstream) requests only. For example mitigations to HTTP/2 flood attacks have been developed for downstream requests only and an upstream server may cause Envoy to consume excessive resources by flooding it with control frames.

However, a number of upcoming Envoy based products are required to be hardened against untrusted components of upstream infrastructure. For example L7ILB proxies requests to customer controlled upstream servers, and a buggy or malicious upstream server in multitenant mode may cause outage or intercept data of other customers.

This design document focuses on hardening against untrusted upstream servers. Hardening against other aspects of upstream infrastructure is covered elsewhere.

### Overview
The entire scope can be split into several areas. This is preliminary list and it may change as we make progress:

- [x] Implement error handling in codecs for cases outside of the processing of incoming client data.
- [x] Achieve parity of security and other bug fixes between downstream and upstream use cases.
- [x] Audit implementation of HCM and Router filters for potential problems with untrusted upstream servers.
- [x] Audit error and invalid response handling in health checking.
- [x] Audit cluster manager.
- [x] Extend fuzzing to handling of upstream responses.

### PR DESCRIPTION
**Pull request by yanavlasov**
Risk Level: Low
Testing: N/A
Docs Changes: Yes
Release Notes: N/A
Platform Specific Features: N/A
Fixes #12278

Signed-off-by: Yan Avlasov <yavlasov@google.com>

### CODE REVIEW

**Review by mattklein123 — DISMISSED** [review]
LGTM assuming we feel we are ready to flip to "enforcing!" Nice work!

**Review by asraa — DISMISSED** [review]
@envoyproxy/security-team

**Review by htuch — DISMISSED** [review]
LGTM. This is a huge milestone, congrats!

File: docs/root/intro/arch_overview/security/threat_model.rst [hunk 1]
```diff
@@ -56,16 +56,11 @@ Data and control plane
...
-In addition, we have an evolving position towards any vulnerability that might be exploitable by
-untrusted upstreams. We recognize that these constitute a serious security consideration, given the
-use of Envoy as an egress proxy. We will activate the security release process for disclosures that
-appear to present a risk profile that is significantly greater than the current Envoy upstream
-hardening status quo.
+these concepts from an architectural perspective. Envoy's core components are considered to be hardened
+against both untrusted downstream and upstream peers. As such our highest priority in risk assessment is the
+threat posed by untrusted downstream client or untrusted upstream server traffic on the data plane. This
+reflects the use of Envoy in an edge serving capacity and also the use of Envoy as a networking component in a
+mesh deployment with unstrusted services.
```
Comment (PiotrSikora) [inline docs/root/intro/arch_overview/security/threat_model.rst:63]: Nit: `s/mesh/service mesh/`

**Review by PiotrSikora — APPROVED** [review]
LGTM sans nit.

File: docs/root/intro/arch_overview/security/threat_model.rst (hunk 1 above)
Comment (yanavlasov) [reply docs/root/intro/arch_overview/security/threat_model.rst]: done

**Review by PiotrSikora — APPROVED** [review]
Thanks!

**Pull request merged**